Add support for running python projects

wmeints opened the pull request:
This PR introduces support for running Python projects as part of an Aspire solution. I took the NodeJS extension as inspiration but had to add a couple of extra pieces of logic since Python is a lot less standardized when it comes to how projects are structured.

Initially I created this project: https://github.com/wmeints/aspire-python but I feel that this PR is a much more useful way to add support for Python. I think it's a great resolution for #2760.

## Assumptions made by the extension

- Your python project has a virtual environment located under `.venv`. Although you can change that.
- You have a Dockerfile for publishing the project to test/acceptance/production environments.

wmeints committed:
Add support for running python projects

wmeints committed:
Add test logic to validate communication to the dashboard

wmeints committed:
Disable https profile for easier testing of python hosting

wmeints committed:
Add support for hosting Flask apps

wmeints committed:
Fix XML Doc

wmeints committed:
Improve code style for tests

mitchdenny commented on Aspire.sln:
```diff
@@ -383,7 +384,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Aspire.Hosting.RabbitMQ", "
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Aspire.Hosting.Redis", "src\Aspire.Hosting.Redis\Aspire.Hosting.Redis.csproj", "{7D57BF77-AD99-48C5-BC4D-AB133A155D97}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Aspire.Hosting.Garnet", "src\Aspire.Hosting.Garnet\Aspire.Hosting.Garnet.csproj", "{96D36588-B223-4381-B774-FB1A3F8BC269}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Aspire.Hosting.Garnet", "src\Aspire.Hosting.Garnet\Aspire.Hosting.Garnet.csproj", "{96D36588-B223-4381-B774-FB1A3F8BC269}"
```
Why did this change?

mitchdenny replied:
It might be a good idea to revert the sln file change and rebase on main and then add the projects back in. I'm not entirely sure why these lines have been updated. But sln file merges can be interesting :)

wmeints replied:
Yeah, that's absolutely true. I don't know what changed that particular line. I used `dotnet sln add` to include the python project. I'll revert the file, rebase, and try again.

wmeints replied:
I redid the change, at looks better now.

mitchdenny commented on playground/python/Python.AppHost/Program.cs:
```diff
@@ -0,0 +1,13 @@
+
+var builder = DistributedApplication.CreateBuilder(args);
+
+builder.AddPythonProjectWithVirtualEnvironment("script-only", "../script_only", "main.py");
```
This is a pretty long `Add...` method. Can we make it more concise?

wmeints replied:
I can remove the virtual environment part and make that implied. I don't think anyone wants to work on python projects without a virtual env.

davidfowl replied:
Make it a bool? Add a doc comment?

mitchdenny commented on src/Aspire.Hosting.Python/FlaskProjectResourceExtensionBuilder.cs:
```diff
@@ -0,0 +1,80 @@
+/// <summary>
+/// Provides extension methods for adding Python Flask applications to an <see cref="IDistributedApplicationBuilder"/>.
+/// </summary>
+public static class FlaskProjectResourceExtensionBuilder
```
I think this could probably collapsed into a single extension class with all the python-related extension methods in it.

wmeints replied:
For the Flask extension that's certainly the case. I'm also working on Django right now, and that's a little more complicated. For Django to run well, you'll need to have two things:

- A super user which you can create with a specific python command
- Applied migrations which requires another command

I would like to make it so that you can use `WithSuperUser(username, secret)` and `WithApplyMigrations()` so that people don't have to fiddle with the commands.

I think I can do this by using a `DjangoProjectResource` and have two extension methods to extend it. What do you think?

mitchdenny replied:
I'm not sure about flask and django at this point. Lets get python + virtual environments sorted out first.

wmeints replied:
Agreed, I'll focus on that first.

mitchdenny commented:
@davidfowl I've been thinking about things like Node.js (and now Python). I'm wondering if whether to make things easier for people we should create a boilerplate Dockerfile if one doesn't already exist. The idea being that it gives people a starting point.

So if someone does `AddPython...`) and points at a particular directory we'll drop down a Dockerfile and .dockerignore file if one doesn't already exist. Then the can tweak it from there customizing the base image etc.

davidfowl commented:
I don't think we should be writing files to disk using these APIs, but I like the idea that they would publish as a docker file by default. The publish operation would throw because it couldn't find one. That would instruct the user to add one for their language.

davidfowl commented on src/Aspire.Hosting.Python/FlaskProjectResourceExtensionBuilder.cs:
```diff
@@ -0,0 +1,83 @@
+        });
+
+        // Expose the port on the FLASK_RUN_PORT to allow the Flask app to bind to the correct port.
+        resourceBuilder.WithHttpEndpoint(targetPort: 5000, name: "http", env: "FLASK_RUN_PORT");
```
Remove the target port.

davidfowl commented on tests/Aspire.Hosting.Tests/Python/AddPythonProjectTests.cs:
```diff
@@ -0,0 +1,54 @@
+
+namespace Aspire.Hosting.Tests.Python;
+
+public class AddPythonProjectTests
```
We need tests that also assert the resource definition and values.

davidfowl commented on src/Aspire.Hosting.Python/PythonProjectResource.cs:
```diff
@@ -0,0 +1,42 @@
+    /// Writes out the publishing manifest for the python project.
+    /// </summary>
+    /// <param name="context">The manifest publishing context to extend.</param>
+    internal async Task WriteDockerFileManifestAsync(ManifestPublishingContext context)
```
How is this different from calling PublishAsDockerFile?

wmeints replied:
It's not, I haven't seen the method ;-) I'll fix it.

wmeints committed:
Clean up playground samples for python

wmeints committed:
Remove flask resource

wmeints committed:
Improve API interface for adding python projects

wmeints committed:
Revert sln file changes

wmeints committed:
Add python hosting project to sln file

wmeints commented:
I can see I broke the build because I forgot to update the playground apphost. I will do that tomorrow morning (GMT+1)

wmeints committed:
Fix playground sample for python

davidfowl commented on tests/Aspire.Hosting.Tests/Python/AddPythonProjectTests.cs:
```diff
@@ -0,0 +1,75 @@
+    private static readonly string s_playgroundDirectory = Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, "../../../../../playground/python/"));
+
+    [LocalOnlyFact("python")]
+    public void AddPythonProject_SetsResourcePropertiesCorrectly()
+    {
+        var pythonProjectDirectory = Path.Combine(s_playgroundDirectory, "script_only");
+        var appBuilder = DistributedApplication.CreateBuilder();
+
+        appBuilder.AddPythonProject("python", pythonProjectDirectory, "main.py");
+
+        var app = appBuilder.Build();
+        var appModel = app.Services.GetRequiredService<DistributedApplicationModel>();
+        var executableResources = appModel.GetExecutableResources();
+
+        var pythonProjectResource = Assert.Single(executableResources);
+
+        Assert.Equal("python", pythonProjectResource.Name);
+        Assert.Equal(pythonProjectDirectory, pythonProjectResource.WorkingDirectory);
+    }
```
I would love to see some pure unit tests like this https://github.com/dotnet/aspire/blob/main/tests/Aspire.Hosting.Tests/Garnet/AddGarnetTests.cs

wmeints replied:
I need a little advice on this. The python virtual env layout varies across platforms, on Windows there's a `scripts` directory while on linux and mac there's a `bin` directory for the executables. I could add OS detection and assume that the scripts/binaries are in the right spot. But I felt that it would be nicer to actually check on disk if the binaries are there.

Because of this choice I figured it would make sense to have local facts with an actual python environment instead of writing pure unit-tests.

I can make purer unit-tests by either: Making the assumption about the virtual env layout and not checking the disk. Or I can write temp dummy files to disk to make the unit-test pass. 

What do you prefer? Or do you have other ideas?

mitchdenny replied:
What I'm doing with the incoming WithDockerfile(...) extension is creating a temp directory and dropping the minimum number of files I need in there. If the tests are going to actually spin up Python that is what you'd need to do.

davidfowl commented on src/Aspire.Hosting.Python/VirtualEnvironment.cs:
```diff
@@ -0,0 +1,50 @@
+    /// <returns>Returns the path to the executable if it exists in the virtual environment.</returns>
+    public string? GetExecutable(string name)
+    {
+        string[] allowedExtensions = [".exe", ".cmd", ".bat", ""];
+        string[] executableDirectories = ["bin", "Scripts"];
```
Filter this based on operating system?

davidfowl commented on src/Aspire.Hosting.Python/PythonProjectResource.cs:
```diff
@@ -0,0 +1,18 @@
+/// <param name="executablePath">The path to the executable used to run the python project.</param>
+/// <param name="projectDirectory">The path to the directory containing the python project.</param>
+public class PythonProjectResource(string name, string executablePath, string projectDirectory)
+    : ExecutableResource(name, executablePath, projectDirectory), IResourceWithServiceDiscovery
```
Do we want this to implement `IResourceWithServiceDiscovery`?

mitchdenny commented on src/Aspire.Hosting.Python/PythonProjectResourceBuilderExtensions.cs:
```diff
@@ -0,0 +1,95 @@
+/// </summary>
+public static class PythonProjectResourceBuilderExtensions
+{
+    /// <summary>
```
The `<summary>` should be a short one sentence thing. Move the rest into `<remarks>` (after `<returns>`).

Would be good to add an `<example>` as well for each of the extension methods.

mitchdenny replied:
@wmeints just wondering if you've had a chance to come back and look at the <remarks> and <example> on this PR?

mitchdenny commented on tests/Aspire.Hosting.Tests/Python/AddPythonProjectTests.cs:
```diff
@@ -0,0 +1,75 @@
+    private static readonly string s_playgroundDirectory = Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, "../../../../../playground/python/"));
+
+    [LocalOnlyFact("python")]
+    public void AddPythonProject_SetsResourcePropertiesCorrectly()
```
@radical heads up on this coming into the repo ... when we start running these tests in C# we are going to need Python available.

mitchdenny commented on tests/Aspire.Hosting.Tests/Python/AddPythonProjectTests.cs:
```diff
@@ -0,0 +1,75 @@
+    [LocalOnlyFact("python")]
+    public void AddPythonProject_SetsResourcePropertiesCorrectly()
+    {
+        var pythonProjectDirectory = Path.Combine(s_playgroundDirectory, "script_only");
```
You are going to need to use the testbuilder here (look at some of the other tests). Otherwise this will leak file watch handles.

davidfowl commented on src/Aspire.Hosting.Python/PythonProjectResourceBuilderExtensions.cs:
```diff
@@ -0,0 +1,95 @@
+            AddProjectArguments(scriptPath, scriptArgs, context);
+        });
+
+        if(!string.IsNullOrEmpty(instrumentationExecutable))
```
nit: Formatting

mitchdenny commented on src/Aspire.Hosting.Python/PythonProjectResourceBuilderExtensions.cs:
```diff
@@ -0,0 +1,95 @@
+    /// <param name="scriptArgs">The arguments for the script.</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<PythonProjectResource> AddPythonProject(
+        this IDistributedApplicationBuilder builder, string name, string projectDirectory, string scriptPath, string virtualEnvironmentPath = ".venv", params string[] scriptArgs)
```
Can you make the default value for `virtualEnvironmentPath` `null` and then check for it and replace it with the default inside the method. Also note the earlier comments I made about making `<summary>` shorter and more concise and adding a `<remarks>` section (where you can have your paragraphs.

ReubenBond commented on src/Aspire.Hosting.Python/PythonProjectResourceBuilderExtensions.cs:
```diff
@@ -0,0 +1,95 @@
+    /// <param name="scriptArgs">The arguments for the script.</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<PythonProjectResource> AddPythonProject(
+        this IDistributedApplicationBuilder builder, string name, string projectDirectory, string scriptPath, string virtualEnvironmentPath = ".venv", params string[] scriptArgs)
```
Should the default `virtualEnvironmentPath` value be taken from the `VIRTUAL_ENV` env var if it's set?

wmeints replied:
@ReubenBond I didn't go for this solution since I expect people to have multiple virtual environments, one for each project. That's at least how I've seen people around me use Aspire.

ReubenBond commented on src/Aspire.Hosting.Python/PythonProjectResourceBuilderExtensions.cs:
```diff
@@ -0,0 +1,95 @@
+            : Path.Join(projectDirectory, virtualEnvironmentPath));
+
+        var instrumentationExecutable = virtualEnvironment.GetExecutable("opentelemetry-instrument");
+        var pythonExecutable = virtualEnvironment.GetRequiredExecutable("python");
```
I wonder how best to handle something like [hatch](https://hatch.pypa.io/latest/environment/#command-execution), which runs projects using the `hatch run ...` command.

wmeints replied:
Good question, I didn't envision running commands other than the regular `python` command. I may add that after this PR is finished.

mitchdenny committed:
Merge branch 'main' into python-hosting-support

mitchdenny commented:
@wmeints this is 95% of the way there.

wmeints commented:
@mitchdenny I've had a few crazy weeks but I haven't forgotten about things! I'm going to take a look at the feedback to address issues. Thanks for reviewing :-)

wmeints committed:
Remove service discovery interface

wmeints committed:
Improve how python executables are resolved

The logic now accounts for operating system so we can make some assumptions about what is and isn't there.

wmeints committed:
Improve test case behavior

wmeints committed:
Improve method docs

wmeints committed:
Add example to extension method

wmeints committed:
Replaced default for virtual env with null

The code now checks for a missing value and replace it with a default
instead of relying on a default as an optional value.

mitchdenny commented on src/Aspire.Hosting.Python/PythonProjectResourceBuilderExtensions.cs:
```diff
@@ -14,21 +14,29 @@ public static class PythonProjectResourceBuilderExtensions
-    /// for the traces, logs, and metrics to be exported to the Aspire observability platform.
-    /// </para>
     /// </summary>
+    /// <example>
```
nit: Can you move the `<example>` below `<remarks>`. In the description for the example could you provide a little bit of detail about the layout of the project to help people understand what "PythonProject" refers to. Like the fact that it is relative to the AppHost directory.

radical commented on tests/Aspire.Hosting.Tests/Python/AddPythonProjectTests.cs:
```diff
@@ -0,0 +1,159 @@
+            Assert.Equal(Path.Join(projectDirectory, ".venv", "bin", "opentelemetry-instrument"), pythonProjectResource.Command);
+        }
+
+        Assert.Equal("--traces_exporter", commandArguments[0]);
+        Assert.Equal("otlp", commandArguments[1]);
+        Assert.Equal("--logs_exporter", commandArguments[2]);
+        Assert.Equal("console,otlp", commandArguments[3]);
+        Assert.Equal("--metrics_exporter", commandArguments[4]);
+        Assert.Equal("otlp", commandArguments[5]);
+        Assert.Equal(pythonExecutable, commandArguments[6]);
+        Assert.Equal(scriptName, commandArguments[7]);
```
nit: This will also check the length, and avoid typos with the indices.

```suggestion
        Assert.Collection(commandArguments,
            arg => Assert.Equal("--traces_exporter", arg),
            arg => Assert.Equal("otlp", arg),
            arg => Assert.Equal("--logs_exporter", arg),
            arg => Assert.Equal("console,otlp", arg),
            arg => Assert.Equal("--metrics_exporter", arg),
            arg => Assert.Equal("otlp", arg),
            arg => Assert.Equal(pythonExecutable, arg),
            arg => Assert.Equal(scriptName, arg));
```

radical commented on tests/Aspire.Hosting.Tests/Python/AddPythonProjectTests.cs:
```diff
@@ -0,0 +1,159 @@
+
+        var commandArguments = await ArgumentEvaluator.GetArgumentListAsync(pythonProjectResource);
+
+        Assert.Equal(scriptName, commandArguments[0]);
```
nit:
```suggestion
        Assert.Collection(commandArguments,
            arg => Assert.Equal(scriptName, arg));
```

radical commented on tests/Aspire.Hosting.Tests/Python/AddPythonProjectTests.cs:
```diff
@@ -0,0 +1,159 @@
+
+            if(instrument)
+            {
+                File.WriteAllText(Path.Join(scriptsDirectory, "opentelemetry-instrument.exe"), "");
```
This path can also be returned from the method, which would simplify the checks for this path.

mitchdenny replied:
Changed this up a bit to support running real Python.

radical commented on tests/Aspire.Hosting.Tests/Python/AddPythonProjectTests.cs:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    [Fact]
+    public async Task AddPythonProjectWithInstrumentation_SwitchesExecutableToInstrumentationExecutable()
```
This should have `instrument: true`, and `instrument: false` cases.

radical commented on src/Aspire.Hosting.Python/PythonProjectResourceBuilderExtensions.cs:
```diff
@@ -0,0 +1,109 @@
+/// <summary>
+/// Provides extension methods for adding Python applications to an <see cref="IDistributedApplicationBuilder"/>.
+/// </summary>
+public static class PythonProjectResourceBuilderExtensions
```
It would be useful to add tests that *run* the python app too. See:
https://github.com/dotnet/aspire/blob/0af8505ff294c4b90ee8cccb5639bf73ec9077ba/tests/Aspire.Hosting.Tests/Node/NodeFunctionalTests.cs#L20-L32

https://github.com/dotnet/aspire/blob/0af8505ff294c4b90ee8cccb5639bf73ec9077ba/tests/Aspire.Hosting.Tests/TestProgramFixture.cs#L83-L92

https://github.com/dotnet/aspire/blob/0af8505ff294c4b90ee8cccb5639bf73ec9077ba/tests/testproject/TestProject.AppHost/TestProgram.cs#L64-L76

radical commented on src/Aspire.Hosting.Python/PythonProjectResourceBuilderExtensions.cs:
```diff
@@ -0,0 +1,109 @@
+/// Provides extension methods for adding Python applications to an <see cref="IDistributedApplicationBuilder"/>.
+/// </summary>
+public static class PythonProjectResourceBuilderExtensions
+{
```
Also, see the `node` related tests in https://github.com/dotnet/aspire/blob/0af8505ff294c4b90ee8cccb5639bf73ec9077ba/tests/Aspire.Hosting.Tests/ManifestGenerationTests.cs#L336-L378

We could have similar for python too.

mitchdenny committed:
Expand remarks and fix up solution file.

mitchdenny commented:
@davidfowl @wmeints so I've been thinking about what `AddPythonProject` does around defaulting to using PublishAsDockerfile. I think that for the initial merge we should leave it to emit an `executable.v0` resource and leave it up to developers to call `PublishAsDockerfile` if they want to.

My reasoning is that I want to add an extension method to `IResourceBuilder<ExecutableResource>` for `WithDockerfile` as the replacement to PublishAsDockerfile.

I think this will lead to less confusion and greater consistency.

mitchdenny committed:
Add manifest generation test.

wmeints committed:
Fix configuration for analyzers

mitchdenny committed:
use real python for testing.

wmeints committed:
Remove UsePublicAnalyzers property from project file

mitchdenny committed:
Merge branch 'python-hosting-support' of https://github.com/wmeints/aspire into python-hosting-support

wmeints commented on src/Aspire.Hosting.Python/Aspire.Hosting.Python.csproj:
```diff
@@ -0,0 +1,29 @@
+
+  <ItemGroup>
+    <AdditionalFiles Include="PublicAPI.Shipped.txt" />
+    <AdditionalFiles Include="PublicAPI.Unshipped.txt" />
```
This was added by VS2022 and I wasn't aware that this was taken care of somewhere else. Removed the extra file entries.

mitchdenny committed:
Merge branch 'python-hosting-support' of https://github.com/wmeints/aspire into python-hosting-support

mitchdenny commented:
@wmeints looks like we are both working on the branch. I switched you project generation code to use real python and actually setup a virtual environment. We need this so we can actually spin up the project in CI to verify things. @radical hopefully we have Python in the environment ;)

wmeints commented:
@mitchdenny, I noticed that you were changing files. I'm going to watch things unfold for a bit. I love how this works on GitHub. I never expected to learn this much about testing a product like Aspire :)

mitchdenny commented:
/azp run

mitchdenny commented:
@radical I think we need to kick the tests in `Aspire.Hosting.Tests` that were recently enabled for CI back to local only until we can fix the issue of the port conflicts.

mitchdenny commented:
/azp run

radical commented on tests/Aspire.Hosting.Tests/Python/AddPythonProjectTests.cs:
```diff
@@ -0,0 +1,296 @@
+
+        var relativePipPath = Path.Combine(
+            ".venv",
+            "Scripts",
```
```suggestion
            OperatingSystem.IsWindows() ? "Scripts" : "bin",
```

radical commented on tests/Aspire.Hosting.Tests/Python/AddPythonProjectTests.cs:
```diff
@@ -0,0 +1,296 @@
+        };
+
+        var createVirtualEnvironmentProcess = Process.Start(prepareVirtualEnvironmentStartInfo);
+        var createVirtualEnvironmentProcessResult = createVirtualEnvironmentProcess!.WaitForExit(10000);
```
It would be very useful to capture the output too, incase the command fails, especially on CI.

radical commented on tests/Aspire.Hosting.Tests/Python/AddPythonProjectTests.cs:
```diff
@@ -0,0 +1,296 @@
+        };
+
+        var installRequirementsProcess = Process.Start(installRequirementsStartInfo);
+        var installRequirementsProcessResult = installRequirementsProcess!.WaitForExit(20000);
```
Should the timeout be longer, for slower CI machines?

mitchdenny committed:
Update tests/Aspire.Hosting.Tests/Python/AddPythonProjectTests.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

mitchdenny committed:
Capturing output.

mitchdenny committed:
Merge branch 'python-hosting-support' of https://github.com/wmeints/aspire into python-hosting-support

mitchdenny committed:
Add overloads.

mitchdenny committed:
Fix test.

mitchdenny commented on src/Aspire.Hosting.Python/PythonProjectResourceBuilderExtensions.cs:
```diff
@@ -0,0 +1,109 @@
+    /// </example>
+    /// <param name="builder">The <see cref="IDistributedApplicationBuilder"/> to add the resource to.</param>
+    /// <param name="name">The name of the resource.</param>
+    /// <param name="projectDirectory">The path to the directory containing the python project files.</param>
```
Covered in remarks.

mitchdenny commented:
/azp run

mitchdenny commented:
/azp run

mitchdenny committed:
Whitespace change.

mitchdenny committed:
Disable failing test which is breaking clean CI (port randomization issue again).

mitchdenny commented:
@wmeints thanks so much for your contribution to .NET Aspire. We'll be shipping your work in .NET Aspire 8.1!

wmeints commented:
Excited to see this appear in 8.1! And thanks again for all the reviews and patience. I learned a lot from this experience. I'm coming back for more 😄

mitchdenny commented:
@wmeints I'm writing up the blog post for .NET Aspire 8.1 are you happy for me to link to your GitHub profile for this contribution?

wmeints commented:
@mitchdenny Feel free to link my profile in the blogpost :-)
